### ISSUE TITLE
BIOS disk interface and other data structures

### ISSUE THREAD

**Issue by ghaerr** [ghaerr/elks#824]
Hello @tkchia,

I thought to move the discussion of how ELKS should work with the BIOS to get/set disk drive parameters here from #823 and #521, since it was getting increasingly off topic.

I have been reading various BIOS source code, as well as looking hard at the ELKS BIOS driver and boot code. The BIOS sources include Award 6.00 source, used in many machines including my own desktop 386, the SeaBios source, used in the QEMU emulator, EMU86 which emulates certain BIOS calls, and the tiny8086 emulator, which provides its BIOS source. All are quite illuminating, although sometimes complicated.

The purpose of this discussion is to better understand how ELKS should interface with a BIOS, with a short term plan to enhance and cleanup the ELKS BIOS driver and the MINIX and FAT boot code, which are all quite messy. With some thought, I think all these can be cleaned up considerably while maintaining known compatibility with older PCs. 

A longer term goal is to understand exactly what external interfaces (BIOS calls, BIOS data areas, special I/O ports, etc) ELKS is reliant on, so that these can be consolidated, to help with a port of ELKS to more modern SBCs, or even another CPU.

The first topic on this subject is "How is the floppy sectors per track (SPT) communicated with the BIOS?"

Increasingly, it is looking like the Diskette Drive Parameter Block (DDPT), sometimes called the Disk Base Table (DBT) may be used to publish BIOS SPT for non-BIOS use, but not to set BIOS SPT. In all BIOS source checked, BIOS INT 13h function 02/03 (read/write sectors) never looks at a "custom DDPT" SPT field to get max sector information. None of the BIOS code ever checks for a rewritten INT 1E, and most never update the DDPT after boot. In several BIOSes, the only time the internal DDPT is accessed is to return it for INT 13h function 08 (get disk parameters).

Thus, the current ELKS "custom DDPT" approach of resetting INT 1Eh and replacing the DDT SPT field should be considered for removal.

It is complicated how each of the BIOSes come up with the SPT to use in the INT 13h read/write calls. Two of them use use the BPB present in a FAT boot sector. For certain drive types, the BIOS will check for validity, then copy the BPB SPT after reading or writing the boot sector (CHS 0/0/1). Thus, it seems that a possible good place to store and communicate SPT is in the ELKS boot sector.

The ELKS boot sector only contains a BPB on FAT boot sectors. I propose to make ELKS boot sectors always compatible with a v4.0 BPB (which we do now for FAT), and store our SPT and other CHS information there, rather than in the ELKS Parameter Block (EPB). Our current BIOS driver probe/open routine always reads the boot sector to get the EPB; this boot sector read would allow those BIOSes that implement the special boot sector handling to update their SPT properly as well.

In conclusion, DDPT seems currently ineffective. The current ELKS BIOS driver attempts reading multiple sectors on a single track for every read, but only for a maximum of two sectors. It has worked without a DDPT for over a year. The current track-cache code just reads from usually different start sector, but also never past the last sector, just like the original driver.

For those BIOSes that don't check CHS in INT 13h, nothing is required, but for those that do, some get that information from the boot sector. ELKS should move to a BPB-compatible boot sector, and avoid CHS duplication. This would have the added benefit, with some validity checking, of fast-probing FAT disks as well.

Our current FAT boot sector has one (!) byte to spare. Removing DDPT there (still keeping FAST_READ) would allow for more space, and insertion of the BPB into the MINIX boot sector would allow for better disk compatibility with other systems as well as ELKS probing.

Thank you!

**Comment by tkchia** [comment]
Hello @ghaerr,

Well, yes --- the point is, I hope to _really_ make sure that ELKS does not need to set the DDPT (or, at least, it can set it less often than I first thought).

As I mentioned, I am concerned about the possibility of ELKS _accidentally_ doing a multitrack read when it does not want one.  If so this will mean ELKS is silently reading erroneous data, which is not good, because (well) correctness is a thing we should probably strive for.

**Comment by tkchia** [comment]
Anyway, as an instance of why I suspect that having an under-counting "sectors per track" field in the DDPT _might_ be problematic: I could not find the Award BIOS sources you mentioned, but I did find a reconstruction of the 11/08/82 IBM PC XT BIOS ([among others](https://sites.google.com/site/pcdosretro/ibmpcbios)).

The XT BIOS very definitely
  * enables "multitrack" I/O when sending read or write commands to the floppy disk controller,
  * reads the "sectors per track" field from the DDPT, and
  * sends the "sectors per track" value to the FDC.

The relevant lines are
```
DISK_READ	PROC	NEAR
	MOV	AL,046H 		; READ COMMAND FOR DMA
J9:					; DISK_READ_CONT
	CALL	DMA_SETUP		; SET UP THE DMA
	MOV	AH,0E6H 		; SET UP RD COMMAND FOR NEC CONTROLLER
	JMP	SHORT RW_OPN		; GO DO THE OPERATION
DISK_READ	ENDP
```
(`0xe6` says to do a read [in multitrack (MT) mode](https://wiki.osdev.org/Floppy_Disk_Controller).  Apparently some older IBM BIOSes have `0x66` here, which means no multitrack mode.)
```
RW_OPN	PROC	NEAR
	...
	MOV	AH,CL			; SECTOR NUMBER
	CALL	NEC_OUTPUT
	MOV	BX,7			; BYTES/SECTOR PARM FROM BLOCK
	CALL	GET_PARM		;  TO THE NEC
	MOV	BX,9			; EOT PARM FROM BLOCK
	CALL	GET_PARM		;  TO THE NEC
	MOV	BX,11			; GAP LENGTH PARM FROM BLOCK
	CALL	GET_PARM		;  TO THE NEC
	...
RW_OPN	ENDP
```
(`EOT` ("end of track") refers to the "sectors per track" field in the DDPT.)
```
GET_PARM	PROC	NEAR
	PUSH	DS			; SAVE SEGMENT
	SUB	AX,AX			; ZERO TO AX
	MOV	DS,AX
	ASSUME	DS:ABS0
	LDS	SI,DISK_POINTER 	; POINT TO BLOCK
	SHR	BX,1			; DIVIDE BX BY 2, AND SET FLAG
					;  FOR EXIT
	MOV	AH,[SI+BX]		; GET THE WORD
	POP	DS			; RESTORE SEGMENT
	ASSUME	DS:DATA
	JC	NEC_OUTPUT		; IF FLAG SET, OUTPUT TO CONTROLLER
	RET				; RETURN TO CALLER
GET_PARM	ENDP
```
Another detail is that the default DDPT in the ROM BIOS says that there can be at most 8 sectors per track, even though the common 360 KiB disk format would have 9 sectors per track:
```
DISK_BASE	LABEL	BYTE
	DB	11001111B	; SRT=C, HD UNLOAD=0F - 1ST SPECIFY BYTE
	DB	2		; HD LOAD=1, MODE=DMA - 2ND SPECIFY BYTE
	DB	MOTOR_WAIT	; WAIT AFTER OPN TIL MOTOR OFF
	DB	2		; 512 BYTES/SECTOR
	DB	8		; EOT (LAST SECTOR ON TRACK)
	DB	02AH		; GAP LENGTH
	DB	0FFH		; DTL
	DB	050H		; GAP LENGTH FOR FORMAT
	DB	0F6H		; FILL BYTE FOR FORMAT
	DB	25		; HEAD SETTLE TIME (MILLISECONDS)
	DB	4		; MOTOR START TIME (1/8 SECONDS)
```

**Comment by tkchia** [comment]
Again, I grant that I might be misunderstanding how "multitrack" reads and writes actually happen on the FDC (the available descriptions are not very detailed).  I also grant that changing the DDPT might not really be needed in most situations (or even at all).  But these are things I would like to make sure first.

Thank you!

**Comment by ghaerr** [comment]
Hello @tkchia,

> I hope to really make sure that ELKS does not need to set the DDPT 
> I am concerned about the possibility of ELKS accidentally doing a multitrack read when it does not want one. 

I completely agree with your thoughts on correctness. Regarding DDPT, I did not mean that it should just be removed, but rather, further inspection should be done to provide evidence either way of what is required to maintain system correctness.
With as many BIOSes as there likely are out there, and their different methods of operation, I felt the need to open this issue so that agreement can be made on a path forward with better understanding.

> I could not find the Award BIOS sources you mentioned

They are [here](https://vetusware.com/download/Award%20BIOS%20v6.00/?id=8399). Create a free login, then download all three .zips. Concatenate them, then unzip the aggregate with "unzip -LL" to create lower case filenames. I created a local git repo so that I could track comments I'm adding, since the source is rather large. I can send you more information on what/where the various things are if you'd like.

The SeaBios sources (very interesting to look at, and well-written in C) are in this repo: `https://git.seabios.org/seabios.git`.

> I did find a reconstruction of the 11/08/82 IBM PC XT BIOS (among others).

Very interesting! Thank you for the link. I will dive into that, as ELKS should definitely support the XT BIOS (even though there aren't too many machines running that anymore).

> The XT BIOS very definitely ... reads the "sectors per track" field from the DDPT

Thank you for the code excerpt!! I would have to say that I stand corrected, as the XT BIOS most definitely uses the DDPT low-memory vector to get the SPT. The question remains as to how the BIOS sets up for 9 SPT: does the BIOS reset its own DDPT with another DISK_BASE, or was it up to MSDOS or another program to rewrite the DDPT vector with that? If the XT BIOS does not rewrite the DDPT, then the ELKS DDPT patching will likely have to stay in.

I'm now guessing, but it seems the BIOS did originally start out requiring use of DDPT, as this was the only method available to patch/upgrade a system to from 320K to 360K (and higher) floppies. And then as BIOSes progressed, the requirement of a patched DDPT no longer was necessary or workable, as they went to greater lengths having to support a wider variety of "floppy" devices which required more information than a DDPT could provide. (As an aside, you will see that the Award BIOS has special support for CDROM and IoMega Zip drives, to the point of rewriting boot sector BPB entries on INT 13h read/write calls).

I need to read the XT BIOS source a bit more, but I would say that given this evidence, the DDPT patching will likely have to stay in, or ELKS would neither boot nor read sectors properly on an original IBM PC with a 360k drive.

That said, what other BIOSes should we support? Probably as many as possible. With DDPT patching in, other BIOSes ignore it, but may still require FAT BPB information in the boot sector, so we should still consider adding a BPB onto MINIX boot disks, and possible ELKS EPB rearrangement. We still don't know exactly how newer BIOSes arrive at the proper SPT other than BPB or CMOS information, since many definitely don't use the DDPT 1Eh vector.

Finally, we need a real IBM PC, or a well-emulated PC with an XT BIOS, for testing.

Thank you!

**Comment by ghaerr** [comment]
Hello @tkchia,

After reading the XT BIOS disk read procedure you provided, I believe I *finally* understand what is going on with how our ELKS original driver has worked *without* a DDPT patch for so long, as well as how the "720k floppy fix" in #39/#44 actually works. I mentioned this in https://github.com/jbruchon/elks/issues/521#issuecomment-717628205 but was not fully correct.

Here is what I think is going on:
- (1) The ELKS BIOS driver uses a 2-sector INT 13h read/write call, after ensuring through CHS calculations it will not be reading past end of track, in which case two seperate INT 13h calls are made.
- (2) The #39/#44 fix for 720k floppies reduces the BIOS INT 13h read/write to a single sector when there are *less than 3* sectors remaining on the track. (This means two sectors remaining, since one sector left would already be a single-sector read).

The reason (1) has been working is that ELKS has been run on either systems with a newer BIOS which don't use DDPT, *or*,  they are running on older systems supporting only smaller (8 or 9 sector) floppies and fix (2) is required.

On older systems, when ELKS is accessing an even-numbered SPT floppy, that means only an 8-sector floppy.
On those older systems, an odd-numbered SPT can be only 360k or 720k floppies, which are the only low-density floppies supported by ELKS.

The ELKS buffer I/O system works with 1KiB logical "blocks", which are sent down to the BIOS block device driver. The driver converts that block to two consecutive physical sectors (possibly not on the same track).

On ELKS, both logical blocks and sector numbers start from 1, not 0. Consider the following on a 9-sector floppy:

When mapping logical blocks to CHS, we find that on odd-number SPT floppies, logical block 5 will map to sectors 9/10. Since 9 is the last sector of the track, fix (2) isn't needed, but the INT 13 call is split into two since the sectors are on seperate tracks. However, on the next track (and every other 2x track multiple), logical sector 9 will map into physical sectors 8/9. Sector 8 is two sectors from the end of track, so fix (2) applies, and splits the call into two INT 13h calls, even though the INT 13h read will be on the same physical track.

The reason that fix (2) is required, *two* sectors from end of track, is exactly the fixed-8-sector DDPT issue in the XT BIOS you provided - the DDPT isn't patched, so the INT 13h read call tries to read two sectors at 8 with a DDPT showing EOT/SPT of 8, and fails! This is exactly the "under-counting" SPT issue you are worried about.

A recent commit (#818) enhanced fix (2) so that it only applies to 9 sector floppies, but now I see that isn't really the whole problem or best fix. Instead, a DDPT patch such as created in #823 (except uncommitted and always ON) will solve that problem, as well as allow the full-track reads to work in the new track-caching scheme. Fix (2) can be removed if a DDPT patch is made for older BIOSes.

I hope this helps, I think we are on the right path, thanks to your information.

Thank you!

**Comment by tkchia** [comment]
Hello @Mellvik,

Would it be possible for you to do a test of what happens if the DDPT's "sectors per track" field is bogotified when loading the kernel?  One way to do this would be to change
```
	mov %al,floppy_table + 4
```
in `bootblocks/boot_sect.S` to, say,
```
	movb $7,floppy_table + 4
```
(and rebuild).  If the "sectors per track" value does have an effect on disk I/O, then this should cause the boot process (for ELKS on a Minix filesystem) to fail in some way --- likely the loaded kernel will hang or the bootloader will error out.  Thank you!

---

Hello @ghaerr,

The Award BIOS seems interesting.  Like the XT BIOS, it
  * enables "multitrack" I/O when sending read or write commands to the floppy disk controller,
  * reads the "sectors per track" field from the DDPT, and
  * sends it to the FDC,

but I see that it has no less than four separate DDPTs, and it uses some sort of probing logic (while loading the boot sector) to decide which DDPT to use at boot time.

SeaBIOS is kind of weird.  It pretty much ignores `int 0x1e` and even its own in-BIOS DDPT, and mainly uses hard-wired values when sending commands to the FDC --- I am not sure why it does this.  At the same time, for each read or write, it actually fakes a "sector per track" value from the starting sector and sector count.  This probably means that ELKS will not accidentally do a "multitrack" read on a SeaBIOS machine (but there might be other issues).

Thank you!

**Comment by tkchia** [comment]
Also, it seems the BIOS does not update the `int 0x1e` vector on its own accord.

`int 0x13` functions `0x08` and `0x18` will return pointers to the appropriate parameter tables, but will not alter `int 0x1e`.  I think the idea is that the OS (or application) is supposed to actually set `int 0x1e` to point to the correct table, or a copy of it, whenever it thinks it is necessary.

(The old XT BIOS implements neither `0x08` and `0x18`.  The Award BIOS implements both.)

Thank you!

**Comment by ghaerr** [comment]
Hello @tkchia,

I missed seeing the part where the Award BIOS fills in "sectors per track" (SPT) from the 1Eh DDPT, thank you.

I downloaded all the IBM BIOS source code versions. Looking at all of them, I find that they all load the SPT from the DDPT using the GET_PARM routine as you pointed out. Interestingly and somewhat strangely though, starting with the [EDIT] XT v1 BIOS versions, multiple internal "DDPT"s are defined, like the Award BIOS, and *some* of their contents are used to program the FDC, but *never* the EOT/SPT or "bytes per sector" fields. Here is the relevant procedure from [EDIT] xtbiosv2/dskette.asm:
```
;----------------------------------------------------------------
; RWV_COM                                                       :
;       THIS ROUTINE SENDS PARAMETERS TO THE NEC SPECIFIC       :
;       TO THE READ/WRITE/VERIFY OPERATIONS.                    :
;                                                               :
; ON ENTRY:     CS:BX = ADDRESS OF MEDIA/DRIVE PARAMETER TABLE  :
; ON EXIT :     @DSKETTE_STATUS, CY REFLECT STATUS OF OPERATION :
;----------------------------------------------------------------
RWV_COM PROC    NEAR
        MOV     AX,OFFSET ER_2          ; LOAD ERROR ADDRESS
        PUSH    AX                      ; PUSH NEC_OUT ERROR RETURN
        MOV     AH,[BP+1]               ; OUTPUT TRACK #
        CALL    NEC_OUTPUT
        MOV     AX,SI                   ; OUTPUT HEAD #
        CALL    NEC_OUTPUT
        MOV     AH,[BP]                 ; OUTPUT SECTOR #
        CALL    NEC_OUTPUT
        MOV     DL,3                    ; BYTES/SECTOR PARAMETER FROM BLOCK
        CALL    GET_PARM                ; . TO THE NEC
        CALL    NEC_OUTPUT              ; OUTPUT TO CONTROLLER
        MOV     DL,4                    ; EOT PARAMETER FROM BLOCK
        CALL    GET_PARM                ; . TO THE NEC
        CALL    NEC_OUTPUT              ; OUTPUT TO CONTROLLER

        MOV     AH,CS:[BX].MD_GAP        ; GET GAP LENGTH

R15:
        CALL    NEC_OUTPUT
        MOV     DL,6                    ; DTL PARAMETER PROM BLOCK
        CALL    GET_PARM                ;  TO THE NEC
        CALL    NEC_OUTPUT              ; OUTPUT TO CONTROLLER
        POP     AX                      ; THROW AWAY ERROR EXIT
ER_2:
        RET
RWV_COM         ENDP
```
Notice in this routine, which is the common code for read/write/verify, CS:BX has the address of one of these internal "DDPT"s, and is used to set the gap length, while the INT 1E DDPT is used to set EOT and bytes/sector.

I am thinking that as support for multiple floppy types were added to the XT BIOS over the years, internal DDPT's were added, but, for compatibility reasons likely with MSDOS or formatting programs, the DDPT 1E vector data was still used for SPT and apparently bytes / sector.

This is strange, because the multi-disk BIOSes we are inspecting do not use the other fields in the INT 1E DDPT. To me, that means that, at the time, a DDPT INT 1E patch to set desired SPT was required before using the INT 13h disk I/O functions, but not otherwise. I am thinking of downloading the MSDOS 6.0 source to check why it patches DDPT.

I would not doubt this continued until the time that LBA was introduced, when the BIOS needed to have its own (unpatched) knowledge of SPT, and started ignoring the DDPT value. This could possibly be confirmed if there is a way to query the BIOS for LBA support.

For ELKS purposes, it is simple to always patch DDPT 1E, after which INT 13h disk I/O should work on all BIOSes.

**Comment by ghaerr** [comment]
Hello @tkchia,

> Also, it seems the BIOS does not update the int 0x1e vector on its own accord.

Yes, I confirm that finding in all BIOSes inspected.

> int 0x13 functions 0x08 and 0x18 will return pointers to the appropriate parameter tables, but will not alter int 0x1e. I think the idea is that the OS (or application) is supposed to actually set int 0x1e to point to the correct table, or a copy of it, whenever it thinks it is necessary.

Interesting. Agreed. I might have to look at what FreeDOS or MSDOS do, out of curiosity.


> (The old XT BIOS implements neither 0x08 and 0x18. The Award BIOS implements both.)

Looking at the IBM source, it seems only AT BIOS v2 started implementing function 0x08, contrary to some [web documentation](https://stanislavs.org/helppc/int_13-8.html) stating "XT and newer".

Do you think it worthwhile to use function 0x08 (get current drive parameters) and use that pointer for the DDPT patch for the BIOS driver, instead of the way I coded it, where a copy of the INT 1Eh DDPT is made and SPT replaced? It doesn't really seem to be necessary, since the more modern BIOSes ignore the DDPT values, and the older multi-disk ones use their own internal copies anyways, for data other than SPT. We would have to, in addition, accurately determine the BIOS version number, which could be a problem.

Thank you!

**Comment by ghaerr** [comment]
Hello @tkchia,

> Looking at the IBM source, it seems only AT BIOS v2 started implementing function 0x08, contrary to some web documentation stating "XT and newer".

Correction: I had an error on my XT BIOS downloads, and only XT v1 was downloaded. It appears that INT 13h function 0x08 was initially implemented in XT BIOS v2 and above, and according to the [version history](https://sites.google.com/site/pcdosretro/ibmbioshist).

So, function 0x08 is "kind of" available on "XT and newer" machines!

Thank you!

**Comment by Mellvik** [comment]
> 
> Would it be possible for you to do a test of what happens if the DDPT's "sectors per track" field is bogotified when loading the kernel? One way to do this would be to change
> 
> 	mov %al,floppy_table + 4
> in bootblocks/boot_sect.S to, say,
> 
> 	movb $7,floppy_table + 4
> (and rebuild). If the "sectors per track" value does have an effect on disk I/O, then this should cause the boot process (for ELKS on a Minix filesystem) to fail in some way --- likely the loaded kernel will hang or the bootloader will error out. Thank you!
> 
> 
> 
@tkchia,
Will take stab at this tomorrow!

-m

**Comment by toncho11** [comment]
Hi,

I do not want to hijack the thread, but it might be related.
I have two IBM 5160:

1) BIOS 1986 -> ELKS boots OK from 360kb floppy disk.
2) BIOS 1983 -> ELKS starts to boot, prints messages but stops with: "VPS: Unable to open root device 0380 SYSTEM HALTED". On this computer there is no HDD controller or HDD attached.

**Comment by ghaerr** [comment]
Hello @toncho11,

> I do not want to hijack the thread, but it might be related.

Thanks for coming to the rescue!

> BIOS 1983 -> ELKS starts to boot, prints messages but stops

Very interesting. Did you run this with the current code, *with* PR #823 included?

The current code, without #823 does not patch a DDPT, so that would likely be the reason.
PR #823, up until just now, had the DDPT removed; I just reenabled it as a result of this discussion and pushed it.
Can you fetch that into your codebase, compile, and try again? The git commands would be:
```
get fetch origin pull/823/head:cache
get checkout cache
make
```

Thank you!

**Comment by ghaerr** [comment]
Hello @tkchia,

> to do this would be to change
 	mov %al,floppy_table + 4
 in bootblocks/boot_sect.S to, say,
 	movb $7,floppy_table + 4
 (and rebuild). 

I tried this on the my Compaq Portable 386: the boot printed "\*" instead of "." every so many sectors... then started the ELKS kernel and crashed.

More information from testing on my Compaq: with your unmodified boot sector, it boots and ELKS runs fine, with or without the DDPT BIOS patch turned on in #823. However, as a result of our discussion above, I have removed the DDPT patch being optional and now it always patches. The original DDPT SPT is now also displayed at boot and it showed 18, which means the Compaq BIOS sets up an accurate INT 1E DDPT, unlike the IBM BIOSes which all have SPT 8.

Since your proposed boot sector change failed on the Compaq, that means that the Compaq BIOS does in fact use the SPT from the patched INT 1E DDPT, just like the source shows the other BIOSes do for INT 13h function 2/3. I've located a disassembled version of the Compaq 386 BIOS [here](https://github.com/jeffpar/pcjs/blob/master/devices/pcx86/rom/compaq/deskpro386/1988-01-28/1988-01-28.asm), and am pretty certain the code that does this is in the procedure starting on line 8257.

The good news: having both the ELKS boot sector as well as the BIOS driver patch the original DDPT with a copy and replacing SPT works on all systems so far, and hopefully it fixes @toncho11's 1983 PC as well.


Thank you!

**Comment by ghaerr** [comment]
Hello @toncho11,

> Can you fetch that into your codebase, compile, and try again? The git commands would be:

Or - if it's easier for you, here is an upload of the 360k ELKS image for you to use:


[fd360-minix.bin.zip](https://github.com/jbruchon/elks/files/5469909/fd360-minix.bin.zip)


Thank you!

**Comment by toncho11** [comment]
I was using commit 60eb6a89 in my previous post.

**Comment by toncho11** [comment]
Hi,

Results of the new build:

1) BIOS 1986 works OK as before
2) BIOS 1983:
- gives same error as before (no hdd attached)
- actually it does load until prompt (but with XT-IDE attached)
      - it says "mounted root 0x0380" and then "failed: /bin/sh"
      - when I type "root" and enter - it says "No shell!" and you can not continue (obviously)
      - also it does not display the "commit" it was built from this time 

So it seems that XT-IDE BIOS helps, but still  even if it says that it mounted root 0x0380 this time  ... it is not possible to use it ...

**Comment by toncho11** [comment]
Hi again,

I just realized that the 1983 BIOS computer had only 256KB of memory. So with the added memory up to 640KB it did boot now without the bash error. But that was with XT-IDE. Then it also worked with an original XT hdd. 

Bottom line: the 1983 works only (it does not give the "VPS: Unable to open root device 0380 SYSTEM HALTED" error) if a hdd drive is attached (XT-IDE or HDD). So this is separate problem, **the most important is that they both are able to complete booting ELKS Linux from commit 2bba6889**.

I also get some strange files like: xx yy zz ... .

**Comment by Mellvik** [comment]
>> Would it be possible for you to do a test of what happens if the DDPT's "sectors per track" field is bogotified when loading the kernel? One way to do this would be to change
>> 
>> 	mov %al,floppy_table + 4
>> in bootblocks/boot_sect.S to, say,
>> 
>> 	movb $7,floppy_table + 4
>> (and rebuild). If the "sectors per track" value does have an effect on disk I/O, then this should cause the boot process (for ELKS on a Minix filesystem) to fail in some way --- likely the loaded kernel will hang or the bootloader will error out. Thank you!
>> 
>> 
>> 
> @tkchia,
> Will take stab at this tomorrow!
> 
> -m
@tkchia,
I'm guessing we're already well past the point where another one of these tests would be useful.

--Mellvik

**Comment by tkchia** [comment]
Hello @ghaerr,

> I tried this on the my Compaq Portable 386: the boot printed "*" instead of "." every so many sectors... then started the ELKS kernel and crashed.

Thanks for the confirmation!  The `*`'s indicate read errors --- which are harmless since the reads can be retried --- but if the kernel crashed, it would indicate that the bootloader was silently reading the wrong things, which was my main worry.

Hello @Mellvik,

> I'm guessing we're already well past the point where another one of these tests would be useful.

Well, thanks for your offer to help though. :-)

**Comment by tkchia** [comment]
Hello @ghaerr,

> Do you think it worthwhile to use function 0x08 (get current drive parameters) and use that pointer for the DDPT patch for the BIOS driver, instead of the way I coded it, where a copy of the INT 1Eh DDPT is made and SPT replaced?

I am not sure this will help very much.

Even if ELKS uses `int 0x13` function `0x08`, it may sometimes still need to make a copy of the DDPT anyway to bump up the "sectors per track" to a high enough value.  One way this may potentially happen, is if ELKS is installed on disks that are formatted by a program akin to [`fdformat`](https://github.com/christoh/fdformat), to squeeze more sectors onto each track than the normal sector count.  (E.g. 21 sectors per track, rather than the normal 18, on a nominally 1.44 MiB diskette.)

Thank you!

**Comment by ghaerr** [comment]
Hello @toncho11,

Great news and thank you for your testing on the 1983 IBM XT. As @tkchia pointed out, a DDPT patch is now proven to be required for older BIOSes. Now, PR #823 not only improves floppy disk performance through caching, but the BIOS driver is bug fixed to operate correctly on older systems using a custom DDPT.

>  the 1983 works only ... if a hdd drive is attached (XT-IDE or HDD). 

Can you send me a screenshot of your boot screen when booting from your 1983 machine without an HD attached, but with 640k memory installed? That will help debug that issue.

> I also get some strange files like: xx yy zz ... .

Those are some test shell scripts I use for network testing, since the binary build was from my home system ELKS tree,.

I really appreciate your jumping in and helping. Your results proved timely and critical to getting this issue resolved!

Thank you!

**Comment by toncho11** [comment]
You are welcomed!

This is with no HDD attached:
![IMG_20201101_181347](https://user-images.githubusercontent.com/1851997/97809741-20aba700-1c6f-11eb-8570-0f6080790373.jpg)

This is with HDD attached:
![IMG_20201101_181559](https://user-images.githubusercontent.com/1851997/97809799-784a1280-1c6f-11eb-96e5-29470d6ae9ee.jpg)

**Comment by ghaerr** [comment]
Hi @toncho11,

OK, I can see what is happening:

The BIOS is using a "Get Drive Parameters" BIOS call (function 8) to ask how many floppy drives are installed, and that BIOS call is returning an error, so it thinks 0.

Later, however, the same Get Drive Parameters BIOS call is made to ask how many hard drives are installed, and that time, the BIOS call doesn't error, and indicates 1.

The BIOS driver currently disables itself if the number of FD and HD drives is 0, that's why it doesn't work.

I will prepare a binary for you to try which should fix this.

Thank you!

**Comment by ghaerr** [comment]
Here @toncho11 is a floppy boot image for you. Can you please try this again both with and without your hard drive, and send screenshots of both boots? It should work in both cases, but I'm interested in seeing what is displayed.

[fd360-minix.bin.zip](https://github.com/jbruchon/elks/files/5471505/fd360-minix.bin.zip)


Thanks!

**Comment by toncho11** [comment]
Hi,

Yes it works! Thanks!

No disk:
![IMG_20201101_200549](https://user-images.githubusercontent.com/1851997/97811930-0200dc80-1c7e-11eb-9f72-22951abc4b9e.jpg)

With real hdd:
![IMG_20201101_201020](https://user-images.githubusercontent.com/1851997/97811960-5efc9280-1c7e-11eb-8549-aed15acdc1dc.jpg)

**Comment by tkchia** [comment]
Hello @ghaerr, hello @Mellvik,

(An interesting side note: I came across this rather old comment in `bioshd.c` --- `git blame` says Riley H. Williams wrote it in 2002 --- which touches on the "sectors per track" field in the DDPT:
```
/*	This is not a bugfix, hence no code, but coders should be aware that
 *	multi-sector reads from this point on depend on bootsect modifying
 *	the default Disk Parameter Block in BIOS.
 *
 *	dpb[4]	should be set to a high value such as 36 so that reads can
 *		go past what is hardwired in the BIOS. 36 is the number of
 *		sectors in a 2.88 floppy track.
 *
 *		If you are booting ELKS with anything other than bootsect,
 *		you have to make equivalent arrangements.
 *
 *	0:0x78	contains the address of dpb (char dpb[12]).
 *
 *	dpb[4]	is the End of Track parameter for the 765 Floppy Disk
 *		Controller.
 *
 *	You may have to copy dpb to RAM as the original is in ROM.
 */
```
I think the general method we are now using to manage and update the DDPT is already more general than what is suggested here.  Most probably we can update --- or simply retire --- this comment in the future.)

Thank you!

**Comment by Mellvik** [comment]
Thank you @tkchia,
Very interesting indeed. And 
I certainly vote for keeping / updating. This type of practical knowledge is hard to come by. Additionally it may - partially or in its entirety - one day become something of an artifact. Like the famous 'you are nor expected to understand this'.

-M

**Comment by toncho11** [comment]
It works on Olivetti M24 also:

![IMG_20201106_160722](https://user-images.githubusercontent.com/1851997/98387882-4b707380-2052-11eb-8cec-5f9d0f1dbee4.jpg)

**Comment by ghaerr** [comment]
@toncho11, you're going to town with ELKS, I love it!

Thanks for the pictures, I'm liking all the ELKS you've got running on your older systems!

BTW, I've got to get you an updated FAT binary, I discovered some other bad problems on large FAT drives, like I fixed for you originally. So be careful with writing on large FAT volumes until you get the update. I'm going to try to write the sys program for you shortly, and I'll send you a binary then.
